### PR TITLE
LF: validate transaction containing Nodes with byKey flag.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This adapates and test the validation of transactions containing nodes
with `byKey` flag.

With respect to #7617, the type of `byKey` is changed from `Boolean` 
to `Option[Boolean]`. The idea is that for old transactions the flag will
be set to `None` (because we don't know if the node was actually done
"by key" or not) while for the next transaction version the flag will
be set to `Some(true)` or `Some(false)` (to be done in an upcoming
PR). This allows us to validate old transaction in a backward
compatible way and check new transaction without injecting the 
serialization version in the node comparison.

This advances the state of #7622

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala [hunk 1]
```diff
@@ -216,7 +213,8 @@ object Node {
       signatories: Set[Party],
       stakeholders: Set[Party],
       key: Option[KeyWithMaintainers[Val]],
-      override val byKey: Boolean // invariant (!byKey || exerciseResult.isDefined)
+      byKey: Option[Boolean] // None if unknown
+      // invariant (!byKey || exerciseResult.isDefined)
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:217]: How about enforcing this invariant with a `require`?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala [hunk 2]
```diff
@@ -362,13 +368,13 @@ object Node {
...
             key2,
-            _,
+            byKey2,
             ) =>
           import nf._
           coid === coid2 && templateId == templateId2 &&
           actingParties.forall(_ => actingParties == actingParties2) &&
           signatories == signatories2 && stakeholders == stakeholders2 &&
-          key.forall(_ => key == key2)
+          key.forall(_ => key == key2) && byKey.forall(_ => byKey == byKey2)
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:377]: Please make sure that you also cover these cases in your other PR #7625 .

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala [hunk 3]
```diff
@@ -34,42 +34,50 @@ private[preprocessing] final class TransactionPreprocessor(
...
+              templateId,
+              exercise.targetCoid,
+              exercise.choiceId,
+              exercise.chosenValue.value,
+            )
+        (cmd, (localCids | newCids.filterNot(globalCids), globalCids))
+      case fetchByKey: Node.NodeFetch[_, _] if fetchByKey.byKey.getOrElse(false) =>
+        val (cmd, newCids) = commandPreprocessor.unsafePreprocessFetchByKey(
+          fetchByKey.templateId,
+          fetchByKey.key.fold(fail("unexpected exerciseByKey without key."))(_.key.value))
```
Comment (andreaslochbihler-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala:73]:
```suggestion
          fetchByKey.key.fold(fail("unexpected fetchByKey without key."))(_.key.value))
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala]: added.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:381]: :+1:

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala]: done.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala [hunk 4]
```diff
@@ -645,12 +657,42 @@ class EngineTest
...
+        reinterpret(
+          engine,
+          Set(alice),
+          tx.roots,
+          tx,
+          txMeta,
+          let,
+          lookupPackage,
+          defaultContracts,
+          defaultKeys)
```
Comment (hurryabit) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:669]: Similar to how we drop the `byKey` flags below, would it make sense to empty the `byKeyNodes` in the metadata for this test?

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala [hunk 5]
```diff
@@ -185,7 +185,7 @@ object TransactionBuilder {
       consuming: Boolean,
       actingParties: Set[String],
       argument: Value,
-      byKey: Boolean = true,
+      byKey: Option[Boolean] = Some(false),
```
Comment (hurryabit) [inline daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:188]: Flipping the default from `true` to `Some(false)` seems strange to me. Having a default value for this at all seems like a giant foot gun anyway. I think we should force the callers to think about how they are exercising.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala [hunk 6]
```diff
@@ -216,9 +213,13 @@ object Node {
       signatories: Set[Party],
       stakeholders: Set[Party],
       key: Option[KeyWithMaintainers[Val]],
-      override val byKey: Boolean // invariant (!byKey || exerciseResult.isDefined)
+      byKey: Option[Boolean] // None if unknown
   ) extends LeafOnlyNode[Cid, Val]
-      with NodeInfo.Fetch
+      with NodeInfo.Fetch {
+
+    require(byKey.forall(!_ || key.isDefined))
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:220]: It took me a moment to grasp this. Maybe this is simpler:
```suggestion
    require(byKey != Some(true) || key.isDefined)
```
Same below.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala [hunk 7]
```diff
@@ -362,13 +372,13 @@ object Node {
...
             key2,
-            _,
+            byKey2,
             ) =>
           import nf._
           coid === coid2 && templateId == templateId2 &&
           actingParties.forall(_ => actingParties == actingParties2) &&
           signatories == signatories2 && stakeholders == stakeholders2 &&
-          key.forall(_ => key == key2)
+          key.forall(_ => key == key2) && byKey.forall(_ => byKey == byKey2)
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:381]: This allows `key == None` and `key2 == Some(_)`. Is that intended? Same for `byKey`. If it is, I'd be more comfortable if we made it more explicit:
```suggestion
          (!key.isDefined() || key == key2) && (!byKey.isDefined() || byKey == byKey2)
```
Same below.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala [hunk 8]
```diff
@@ -283,6 +283,13 @@ sealed abstract class HasTxNodes[Nid, +Cid, +Val] {
 
   def roots: ImmArray[Nid]
 
+  def byKeyNodes: Set[Nid] =
+    nodes.collect {
+      case (nodeId, node: Node.NodeExercises[_, _, _]) if node.byKey.getOrElse(false) => nodeId
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:288]: Does this really do what we want? Or do we need
```suggestion
      case (nodeId, node): (_, Node.NodeExercises[_, _, _]) if node.byKey.getOrElse(false) => nodeId
```
There is [this related forum post](https://discuss.daml.com/t/v-ty-in-a-pattern-is-not-an-annotation-scala-ritual/1450) from @S11001001, which really scares me.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionSpec.scala [hunk 9]
```diff
@@ -234,7 +234,7 @@ object TransactionSpec {
       children = children,
       exerciseResult = if (hasExerciseResult) Some(V.ValueUnit) else None,
       key = None,
-      byKey = false
+      byKey = Some(false)
```
Comment (hurryabit) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionSpec.scala:237]: Do we need a new test for the `byKey = None` case as well?

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoDivulgenceSpec.scala [hunk 10]
```diff
@@ -68,7 +68,7 @@ private[dao] trait JdbcLedgerDaoDivulgenceSpec extends LoneElement with Inside {
           children = ImmArray.empty,
           exerciseResult = None,
           key = None,
-          byKey = false,
+          byKey = None,
```
Comment (hurryabit) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoDivulgenceSpec.scala:71]: Do we need to test the `byKey = Some(_)` cases as well?

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoSuite.scala [hunk 11]
```diff
@@ -129,7 +129,7 @@ private[dao] trait JdbcLedgerDaoSuite extends AkkaBeforeAndAfterAll with JdbcLed
       children = ImmArray.empty,
       exerciseResult = Some(ValueText("some exercise result")),
       key = None,
-      byKey = false,
+      byKey = None,
```
Comment (hurryabit) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoSuite.scala:132]: Do we need to test the `byKey = Some(_)` cases as well?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ProjectionsSpec.scala [hunk 12]
```diff
@@ -51,7 +51,7 @@ class ProjectionsSpec extends WordSpec with Matchers {
       children = ImmArray.empty,
       exerciseResult = None,
       key = None,
-      byKey = false,
+      byKey = None,
```
Comment (hurryabit) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ProjectionsSpec.scala:54]: Do we need a test for `byKey = Some(false)` as well?

**Review by hurryabit — APPROVED** [review]
This looks very fine to me. Thanks a lot. And apologies for the repetitive questions about tests.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala (hunk 8 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:288]: We are actually doing pattern matching, so it is fine to use my initial code. Concerning @S11001001's I will argue you should avoid using pattern matching for initializing value. 

Your code is not working anyway, because of type erasure, i.e. the type `(_, Node.NodeExercises[_, _, _])` becomes at compile time `(_, _)`. 

By tricking a bit you proposal I manage to please the compiler with:
```suggestion
       case tuple: (_, Node.NodeExercises[_, _, _]) if tuple._1.byKey.getOrElse(false) => tuple._2
```
Note it however emit the "nice" warning: 
```
.../Transaction.scala:288: warning: non-variable type argument com.daml.lf.transaction.Node.NodeExercises[_,_,_] in type pattern (_, com.daml.lf.transaction.Node.NodeExercises[_,_,_]) is unchecked since it is eliminated by erasure
      case tuple: (_, Node.NodeExercises[_, _, _]) if tuple._1.byKey.getOrElse(false) => tuple._2
```

I think this should be an error not a warning.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala (hunk 4 above)
Comment (hurryabit) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:669]: I see. Thank you.

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala (hunk 5 above)
Comment (hurryabit) [reply daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:188]: Thank you.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala (hunk 8 above)
Comment (hurryabit) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:288]: This is so tricky. Thanks for your explanation.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 6 above)
Comment (S11001001) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:220]: But strictly less safe; you will not get a type error if `byKey`'s type changes, whereas the prior form compiles iff `byKey` has a `forall` method and its typearg has a `!` method that returns a value with a `||` method that returns the result expected by `forall`.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala (hunk 8 above)
Comment (S11001001) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:288]: The prior code is the proper way to express this, because indeed a pattern is what you want here. The advice about placement to get an annotation from my post doesn't apply here, because there is nowhere you can put an annotation in a `case` anyway.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 6 above)
Comment (hurryabit) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:220]: Why would it not fail with a type error? Is the type of `!=` too general or would it use `!=` at the least common supertype of `byKey`'s type and `Option[Boolean]`?

Comment (andreaslochbihler-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:220]: How about `!byKey.contains(true) || key.isDefined)`?

Comment (S11001001) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:220]: @hurryabit The former; if you want any safety you must either explicitly type both arguments or use an alternative equality operator from a library.

Comment (hurryabit) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:220]: @S11001001 This is sad. 😿
@remyhaemmerle-da Let's go with @andreaslochbihler-da's suggestion then.

**Comment by cocreature** [comment]
@remyhaemmerle-da is this obsoleted by the recent `byKey` changes?

**Comment by cocreature** [comment]
@remyhaemmerle-da ping can we close this or not?

**Comment by remyhaemmerle-da** [comment]
Need to much work to be rebase.
